### ISSUE TITLE
Infer function argument types from decorator

### ISSUE THREAD

**Issue by schiegl** [microsoft/pyright#6905]
I saw that pyright has some [features](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#call-site-return-type-inference) to infer types of function arguments when they're called. But it doesn't seem to be able to infer the types in this example.

```py
from typing import Callable

def decorator(fun: Callable[[int], int]) -> Callable[[int], int]:
    def wrapper(a: int) -> int:
        return fun(a)

    return wrapper

@decorator
def myfun(a):
    return a
```

From this example it's clear that `a` must be an `int` if I'm not mistaken. And in fact, when trying to call `myfun(` pyright correctly suggests that the argument `a` should be an int. However, when hovering `a` through the language server the type of `a` is `Unknown`. Also, inside `myfun` there is no information known about `a`.

In this example I'm trying to use decorator as a sort of abstract class e.g. making the user of `@decorator` conform to what is expected.

**Comment by erictraut** [comment]
It's unlikely that we'd ever implement this. It would require some form of "reverse inference", which would be highly inaccurate (many false positives).

**Comment by tahsmith** [comment]
How is this different significantly different to lambda parameters?

E.g., the following inference is supported:

`f(lambda x: x)` x can be inferred if f takes, for example,`Callable[[int], int]`

Also, I had been looking to implement this myself before I found this question, since I have two good use cases for this:

1. An `implements` decorator to declare a function matches a given signature. Would remove boiler plate if types don’t have to be repeated.
2. Allowing pyright to show better suggestions for `hypothesis` if the types could be inferred from the usage of `given` decorator. 

Both of these would only need the same inference capability that already exists for lambdas.
